src/Sulu/Component/HttpCache/Handler/TagsHandler.php

```diff
@@ -106,16 +108,51 @@ public function flush()
             return false;
         }
 
+        $tags = [];
         foreach ($this->structuresToInvalidate as $structure) {
-            $banKey = $this->getBanKey($structure->getUuid());
+            $tags[] = $this->getTagKey($structure->getUuid());
+        }
 
+        $this->doInvalidate($tags);
+        $this->proxyClient->flush();
+    }
+
+    /**
+     * TODO: In the future the FOSHttpClient Varnish proxy will implement the TagInterface and this
+     *       code will no longer be necessary.
+     */
+    private function doInvalidate(array $tags)
+    {
+        if ($this->proxyClient instanceof TagInterface) {
+            $this->proxyClient->invalidateTags($tags);
+
+            return;
+        }
+
+        if ($this->proxyClient instanceof BanInterface) {
+            $tagExpression = sprintf('(%s)(,.+)?$', implode('|', array_map('preg_quote', $this->escapeTags($tags))));
             $this->proxyClient->ban([
-                self::TAGS_HEADER => sprintf('(%s)(,.+)?$', preg_quote($banKey)),
+                self::TAGS_HEADER => $tagExpression,
             ]);
+
+            return;
         }
 
-        $this->proxyClient->flush();
+        throw new \RuntimeException(sprintf(
+            'Proxy client must either support BAN (BanInterface) or tag invalidation (TagInterface), "%s" supports neither',
+            get_class($this->proxyClient)
+        ));
+    }
+
+    /**
+     * Escape tags (should be handled by the client, see comment on `doInvalidate`.
+     */
+    private function escapeTags(array $tags)
+    {
+        array_walk($tags, function (&$tag) {
+            $tag = str_replace([',', "\n"], ['_', '_'], $tag);
```

Comment:
I wonder if we should just throw an exception if it contains illegal characters e.g. anything out side of chars 0-127 would be illegal? We cache UUIDs so this would be safe